### ISSUE TITLE
noise generator code 

### ISSUE THREAD

**Issue by zj19921221** [bethgelab/game-of-noise#1]
I am confused about the how to generator noise by training a tiny CNN when i read the paper ; 
how to joint training with the classify CNN?
Can you release code of joint training ?

thanks a lot

**Comment by EvgeniaAR** [comment]
Sure :). This is the noise generator class:

    class NoiseResNet1x1Conv(nn.Module):
        def __init__(self, channels, custom_init=True):
            super().__init__()
            self.conv_2d_1 = nn.Conv2d(in_channels=channels, out_channels=20, kernel_size=1, stride=1,
                                    padding=0)
            self.conv_2d_2 = nn.Conv2d(in_channels=20, out_channels=20, kernel_size=1, stride=1,
                                    padding=0)
            self.conv_2d_3 = nn.Conv2d(in_channels=20, out_channels=20, kernel_size=1, stride=1,
                                    padding=0)
            self.conv_2d_4 = nn.Conv2d(in_channels=20, out_channels=channels, kernel_size=1, stride=1,
                                    padding=0)
            if custom_init:
                for m in self.modules():
                    if isinstance(m, nn.Conv2d):
                        m.weight.data.normal_(0.0, 0.01)
                        m.bias.data.zero_()

        def forward(self, x):
            bs, ch, nx, ny = x.shape
            x = torch.empty((bs, ch, nx, ny), device=x.device).normal_()

            residual = x
            x = F.relu(self.conv_2d_1(x))
            x = F.relu(self.conv_2d_2(x))
            x = F.relu(self.conv_2d_3(x))
            x = self.conv_2d_4(x) + residual

            return x

The class can be instantiated as:

    noise_gen = NoiseResNet1x1Conv(channels=3, custom_init=True)

**Comment by zj19921221** [comment]
I am glad to receive your reply; 
I read the paper several times.  In my opinion, joint training is like the training of GAN; the loss function is nearly same as the GAN；
But I am confused about combining the adversarial noise with the normal noise augmentation when training the net alike GAN；
what I really need is the script of joint_train；

look forward your reply!!!

**Comment by EvgeniaAR** [comment]
We are working on releasing the full training code. It might take us some time to clean and comment it properly, but we are on it! :)

**Comment by zj19921221** [comment]
thanks a lot;  look forward the time you release the full script;

**Comment by EvgeniaAR** [comment]
The training code is online. Please tell me of any bugs you encounter.

**Comment by zj19921221** [comment]
> The training code is online. Please tell me of any bugs you encounter.

thanks a lot; thank you !!!

**Comment by zj19921221** [comment]
hi EvgeniaAR I have read the code today morning; I found a  difference, the way of dealing with the noise and original img; In my previous experiments, I just add the noise with the original image;
but i found you use the function "generate_noisy_data" to complete this block; I am confused of this;?

**Comment by EvgeniaAR** [comment]
hi there. we essentially do just that (add the noise to the original image), but we wrote a function for it. The function generate_noisy_data does several things:

1. one has to specify what type of noise one wants -> Gaussian, adversarial, speckle...
then, generate_noisy_data has several if-else statements for each noise option.
2. one samples the specified noise and adds it to the original image
3. in case of adversarial noise, we also sample noise from previous noise generators to do experience replay.

**Comment by annopackage** [comment]
Hi, I am a little confused about the function 'fix_perturbation_size' and 'zero_accuracy_perturbation' in evaluation_utils.py? What's the usage of them?

**Comment by EvgeniaAR** [comment]
Hi there. 1. The function 'fix_perturbation_size' forces the added noise to have a specific perturbation size after clipping. The problem is if I fix the norm of Gaussian noise, add it and clip it -> the norm will become smaller after clipping. This function is explained shortly in our revised paper version on page 6, footnote 2 and in more detail in "Jonas Rauber and Matthias Bethge. Fast differentiable clipping-aware normalization and rescaling. arXiv preprint arXiv:2007.07677, 2020, https://github.com/
jonasrauber/clipping-aware-rescaling."
2. The function 'zero_accuracy_perturbation' takes a noise pattern, e.g. Gaussian or uniform, and performs a binary search of how much of this noise must be added in order to get to the decision boundary. So essentially, we measure the distance to the decision boundary (in the L2 norm) if the direction is sampled normally/ uniformly/ adversarially. Side notes: 1. We sample the noise once for every target image in the mini-batch, 2. We calculate this distance vor every image in the test set and report the median.
Do the answers clarify your questions?

**Comment by annopackage** [comment]
Thanks for your reply which clarifies most of my confusion. However, I also have another two questions:
1. Does the zero_accuracy_perturbation serve as function to find the boundary: epsilon? Why do we have to find this boundary?
2.  How to pre-define the epsilon while training in 'ANT+SIN' mode, such as select 135 as default?

**Comment by EvgeniaAR** [comment]
1. a) Yes. b) Well, the reasoning is that we have to somehow quantify how severe a particular noise pattern is. Is our adversarial noise more successful in fooling the classifier compared to Gaussian noise? To gauge this, we calculate the distance to the decision boundary, epsilon.
2. An epsilon of 135 corresponds to the perturbation size of Gaussian noise with sigma=0.5 which happened to be the optimal sigma in order to enhance corruption robustness with Gaussian data augmentation. We have tried other epsilons as well and it did not make much of a difference. We even tried varying epsilon by keeping the accuracy of the classifier fixed during training which also worked. In the latter case, for each mini-batch, we fixed the perturbation size of added noise so that around 30% of the images were misclassified. In this case, as the classifier got more and more robust, the perturbation size kept increasing.

**Comment by annopackage** [comment]
Your answer has solved my issue. Thanks a lot. BTW, what's the custom training procedure in adversarial training?

**Comment by EvgeniaAR** [comment]
sorry,I don't understand your question. What do you mean with custom training procedure?

**Comment by annopackage** [comment]
Hi, I mean the detailed procedure used in your paper.

**Comment by EvgeniaAR** [comment]
Our training procedure is explained in section 3 (methods) of our paper (https://arxiv.org/abs/2001.06057). The relevant hyper-parameters and other training details can be found in Appendix A+B.

**Comment by annopackage** [comment]
ok, thanks a lot.
